### PR TITLE
Add new aten::device variant to TorchScript

### LINKED ISSUE #96627: Pytorch 2 Compile Doesn't with T5 Models + OnnxRT + Cuda
### 🐛 Describe the bug

Pytorch Compile using OnnxRt backend works with T5 models only on CPU but not on Cuda.

Complete reproducible results could be seen here:
https://colab.research.google.com/drive/1xSRiz91hNTCDkiuMnsQym1bYYb1atmCj?usp=sharing

### Error logs

```
/usr/local/lib/python3.9/dist-packages/torch/jit/_check.py:172: UserWarning: The TorchScript type system doesn't support instance-level annotations on empty non-base types in `__init__`. Instead, either 1) use a type annotation in the class body, or 2) wrap the type in `torch.jit.Attribute`.
  warnings.warn("The TorchScript type system doesn't support "
======= Diagnostic Run torch.onnx.export version 2.1.0.dev20230312+cu118 =======
verbose: False, log level: Level.ERROR
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

/usr/local/lib/python3.9/dist-packages/torch/jit/_check.py:172: UserWarning: The TorchScript type system doesn't support instance-level annotations on empty non-base types in `__init__`. Instead, either 1) use a type annotation in the class body, or 2) wrap the type in `torch.jit.Attribute`.
  warnings.warn("The TorchScript type system doesn't support "
---------------------------------------------------------------------------
BackendCompilerFailed                     Traceback (most recent call last)
[<ipython-input-7-8d0c079efa2f>](https://localhost:8080/#) in <module>
      1 with torch.no_grad():
----> 2   output = model(**{
      3       "input_ids": ids["input_ids"],
      4       "attention_mask": ids["attention_mask"],
      5       "decoder_input_ids": ids["input_ids"],

31 frames
[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py](https://localhost:8080/#) in __call__(self, *args, **kwargs)
    116 
    117     def __call__(self, *args, **kwargs):
--> 118         return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
    119 
    120     def forward(self, *args, **kwargs):

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py](https://localhost:8080/#) in _fn(*args, **kwargs)
    252             dynamic_ctx.__enter__()
    253             try:
--> 254                 return fn(*args, **kwargs)
    255             finally:
    256                 set_eval_frame(prior)

[/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *args, **kwargs)
   1499                 or _global_backward_pre_hooks or _global_backward_hooks
   1500                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501             return forward_call(*args, **kwargs)
   1502         # Do not call functions when jit is used
   1503         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.9/dist-packages/transformers/models/t5/modeling_t5.py](https://localhost:8080/#) in forward(self, input_ids, attention_mask, decoder_input_ids, decoder_attention_mask, head_mask, decoder_head_mask, cross_attn_head_mask, encoder_outputs, past_key_values, inputs_embeds, decoder_inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict)
   1624         if encoder_outputs is None:
   1625             # Convert encoder inputs in embeddings if needed
-> 1626             encoder_outputs = self.encoder(
   1627                 input_ids=input_ids,
   1628                 attention_mask=attention_mask,

[/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *args, **kwargs)
   1499                 or _global_backward_pre_hooks or _global_backward_hooks
   1500                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501             return forward_call(*args, **kwargs)
   1502         # Do not call functions when jit is used
   1503         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.9/dist-packages/transformers/models/t5/modeling_t5.py](https://localhost:8080/#) in forward(self, input_ids, attention_mask, encoder_hidden_states, encoder_attention_mask, inputs_embeds, head_mask, cross_attn_head_mask, past_key_values, use_cache, output_attentions, output_hidden_states, return_dict)
    978         # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
    979         # ourselves in which case we just need to make it broadcastable to all heads.
--> 980         extended_attention_mask = self.get_extended_attention_mask(attention_mask, input_shape)
    981 
    982         # If a 2D or 3D attention mask is provided for the cross-attention

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py](https://localhost:8080/#) in catch_errors(frame, cache_size)
    389 
    390         with compile_lock:
--> 391             return callback(frame, cache_size, hooks)
    392 
    393     catch_errors._torchdynamo_orig_callable = callback  # type: ignore[attr-defined]

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py](https://localhost:8080/#) in _convert_frame(frame, cache_size, hooks)
    404         counters["frames"]["total"] += 1
    405         try:
--> 406             result = inner_convert(frame, cache_size, hooks)
    407             counters["frames"]["ok"] += 1
    408             return result

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py](https://localhost:8080/#) in _fn(*args, **kwargs)
    103         torch.fx.graph_module._forward_from_src = fx_forward_from_src_skip_result
    104         try:
--> 105             return fn(*args, **kwargs)
    106         finally:
    107             torch._C._set_grad_enabled(prior_grad_mode)

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py](https://localhost:8080/#) in _convert_frame_assert(frame, cache_size, hooks)
    261         initial_grad_state = torch.is_grad_enabled()
    262 
--> 263         return _compile(
    264             frame.f_code,
    265             frame.f_globals,

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py](https://localhost:8080/#) in time_wrapper(*args, **kwargs)
    162                 compilation_metrics[key] = []
    163             t0 = time.time()
--> 164             r = func(*args, **kwargs)
    165             time_spent = time.time() - t0
    166             # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py](https://localhost:8080/#) in _compile(code, globals, locals, builtins, compiler_fn, one_graph, export, hooks, frame)
    324         for attempt in itertools.count():
    325             try:
--> 326                 out_code = transform_code_object(code, transform)
    327                 orig_code_map[out_code] = code
    328                 break

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/bytecode_transformation.py](https://localhost:8080/#) in transform_code_object(code, transformations, safe)
    528     propagate_line_nums(instructions)
    529 
--> 530     transformations(instructions, code_options)
    531     return clean_and_assemble_instructions(instructions, keys, code_options)[1]
    532 

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py](https://localhost:8080/#) in transform(instructions, code_options)
    311         )
    312         with tracing(tracer.output.tracing_context):
--> 313             tracer.run()
    314         output = tracer.output
    315         assert output is not None

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py](https://localhost:8080/#) in run(self)
   1839     def run(self):
   1840         _step_logger()(logging.INFO, f"torchdynamo start tracing {self.f_code.co_name}")
-> 1841         super().run()
   1842 
   1843     def match_nested_cell(self, name, cell):

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py](https://localhost:8080/#) in run(self)
    595                     self.instruction_pointer is not None
    596                     and not self.output.should_exit
--> 597                     and self.step()
    598                 ):
    599                     pass

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py](https://localhost:8080/#) in step(self)
    558             if not hasattr(self, inst.opname):
    559                 unimplemented(f"missing: {inst.opname}")
--> 560             getattr(self, inst.opname)(inst)
    561 
    562             return inst.opname != "RETURN_VALUE"

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py](https://localhost:8080/#) in RETURN_VALUE(self, inst)
   1918         )
   1919         log.debug("RETURN_VALUE triggered compile")
-> 1920         self.output.compile_subgraph(
   1921             self, reason=GraphCompileReason("return_value", [self.frame_summary()])
   1922         )

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py](https://localhost:8080/#) in compile_subgraph(self, tx, partial_convert, reason)
    567             if count_calls(self.graph) != 0 or len(pass2.graph_outputs) != 0:
    568                 output.extend(
--> 569                     self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
    570                 )
    571 

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py](https://localhost:8080/#) in compile_and_call_fx_graph(self, tx, rv, root)
    613 
    614         assert_no_fake_params_or_buffers(gm)
--> 615         compiled_fn = self.call_user_compiler(gm)
    616         compiled_fn = disable(compiled_fn)
    617 

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py](https://localhost:8080/#) in time_wrapper(*args, **kwargs)
    162                 compilation_metrics[key] = []
    163             t0 = time.time()
--> 164             r = func(*args, **kwargs)
    165             time_spent = time.time() - t0
    166             # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py](https://localhost:8080/#) in call_user_compiler(self, gm)
    699             assert callable(compiled_fn), "compiler_fn did not return callable"
    700         except Exception as e:
--> 701             raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
    702                 e.__traceback__
    703             ) from None

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py](https://localhost:8080/#) in call_user_compiler(self, gm)
    695                 compiled_fn = compiler_fn(gm, self.example_inputs())
    696             else:
--> 697                 compiled_fn = compiler_fn(gm, self.fake_example_inputs())
    698             _step_logger()(logging.INFO, f"done compiler function {name}")
    699             assert callable(compiled_fn), "compiler_fn did not return callable"

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/debug_utils.py](https://localhost:8080/#) in debug_wrapper(gm, example_inputs, **kwargs)
   1062                     raise
   1063         else:
-> 1064             compiled_gm = compiler_fn(gm, example_inputs)
   1065 
   1066         return compiled_gm

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/backends/common.py](https://localhost:8080/#) in wrapper(model, inputs, **kwargs)
    105         with _disable_current_modes():
    106             inputs = list(map(defake, inputs))
--> 107             return fn(model, inputs, **kwargs)
    108 
    109     return wrapper

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/backends/onnxrt.py](https://localhost:8080/#) in onnxrt(gm, example_inputs, filename, provider)
     53     if filename is None:
     54         with tempfile.NamedTemporaryFile(suffix=".onnx") as tmp:
---> 55             return onnxrt(gm, example_inputs, filename=tmp.name)
     56 
     57     import onnxruntime  # type: ignore[import]

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/backends/common.py](https://localhost:8080/#) in wrapper(model, inputs, **kwargs)
    105         with _disable_current_modes():
    106             inputs = list(map(defake, inputs))
--> 107             return fn(model, inputs, **kwargs)
    108 
    109     return wrapper

[/usr/local/lib/python3.9/dist-packages/torch/_dynamo/backends/onnxrt.py](https://localhost:8080/#) in onnxrt(gm, example_inputs, filename, provider)
     68 
     69     torch.onnx.export(
---> 70         torch.jit.script(gm),
     71         example_inputs,
     72         filename,

[/usr/local/lib/python3.9/dist-packages/torch/jit/_script.py](https://localhost:8080/#) in script(obj, optimize, _frames_up, _rcb, example_inputs)
   1282     if isinstance(obj, torch.nn.Module):
   1283         obj = call_prepare_scriptable_func(obj)
-> 1284         return torch.jit._recursive.create_script_module(
   1285             obj, torch.jit._recursive.infer_methods_to_compile
   1286         )

[/usr/local/lib/python3.9/dist-packages/torch/jit/_recursive.py](https://localhost:8080/#) in create_script_module(nn_module, stubs_fn, share_types, is_tracing)
    478     if not is_tracing:
    479         AttributeTypeIsSupportedChecker().check(nn_module)
--> 480     return create_script_module_impl(nn_module, concrete_type, stubs_fn)
    481 
    482 def create_script_module_impl(nn_module, concrete_type, stubs_fn):

[/usr/local/lib/python3.9/dist-packages/torch/jit/_recursive.py](https://localhost:8080/#) in create_script_module_impl(nn_module, concrete_type, stubs_fn)
    544     # Compile methods if necessary
    545     if concrete_type not in concrete_type_store.methods_compiled:
--> 546         create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
    547         # Create hooks after methods to ensure no name collisions between hooks and methods.
    548         # If done before, hooks can overshadow methods that aren't exported.

[/usr/local/lib/python3.9/dist-packages/torch/jit/_recursive.py](https://localhost:8080/#) in create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
    395     property_rcbs = [p.resolution_callback for p in property_stubs]
    396 
--> 397     concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
    398 
    399 def create_hooks_from_stubs(concrete_type, hook_stubs, pre_hook_stubs):

BackendCompilerFailed: backend='onnxrt' raised:
RuntimeError: 
Arguments for call are not valid.
The following variants are available:
  
  aten::device(str a) -> Device:
  Argument a not provided.
  
  device(str type) -> Device:
  Keyword argument index unknown.

The original call is:
  File "<eval_with_key>.3", line 25
    transpose_3 = transpose_1.transpose(3, 2);  transpose_1 = None
    matmul = torch.matmul(transpose, transpose_3);  transpose = transpose_3 = None
    arange = torch.arange(6, dtype = torch.int64, device = device(type='cuda', index=0))
                                                           ~~~~~~ <--- HERE
    getitem = arange[(slice(None, None, None), None)];  arange = None
    arange_1 = torch.arange(6, dtype = torch.int64, device = device(type='cuda', index=0))



You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

### Minified repro

_No response_

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230312+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.9.16 (main, Dec  7 2022, 01:11:51)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.147+-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.85.12
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:                        3
CPU MHz:                         2000.138
BogoMIPS:                        4000.27
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        1 MiB
L3 cache:                        38.5 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+2c32f43999
[pip3] torch==2.1.0.dev20230312+cu118
[pip3] torchaudio==0.13.1+cu116
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.14.1
[pip3] torchvision==0.14.1+cu116
[conda] Could not collect
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes #96627 

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by davidberard98** [comment]
Feel free to add me as a reviewer when you need the PR to be reviewed.

**Comment by thiagocrepaldi** [comment]
> Feel free to add me as a reviewer when you need the PR to be reviewed.

Thank you. The current draft tries to add a variant to `aten::device`, but apparently something is missing because I got the exact same error. Any clues?

File: torch/csrc/jit/mobile/promoted_prim_ops.cpp
Comment (davidberard98) [inline torch/csrc/jit/mobile/promoted_prim_ops.cpp:101]:
```suggestion
  at::Tensor tensor;
  int index;
  pop(stack, tensor);
  pop(stack, index);
  auto device = tensor.device();
  device.set_index(index);
  push(stack, device);
```

**Comment by davidberard98** [comment]
^^ not sure if that's the issue. we can wait for CI to see if you recent changes fixed the previous errors in CI.

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:497]: actually, I think you might want to instead duplicate the `aten::device` op around line ~2200

Comment (thiagocrepaldi) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:497]: Ouch, I've added a variant for `prim::device`, not `aten::device`. Thanks for catching that
That even changes things, because `aten::device`'s inputs will be `str, int`, not `Tensor, int`

File: torch/csrc/jit/mobile/promoted_prim_ops.cpp
Comment (thiagocrepaldi) [reply torch/csrc/jit/mobile/promoted_prim_ops.cpp:101]: Thanks, will have to refactor so that inputs are `string, int` instead of `Tensor, int`

**Comment by kit1980** [comment]
To resolve the CLA issue, make sure to add the email address of the commit to your GitHub profile
(to see the actual email address, add .patch at the end of this PR page's URL).

**Review by jgong5 — APPROVED** [review]
would you please add unit test?

**Comment by thiagocrepaldi** [comment]
> **pull / linux-focal-py3-clang7-mobile-build / build (pull_request) ** Failing after 8m

Sure, I can add unit test when I find out what is going on with the `pull / linux-focal-py3-clang7-mobile-build / build (pull_request) Failing after 8m` and `pull / linux-focal-py3-clang7-mobile-custom-build-static / build (pull_request)`

It complains about `aten::device` being already registered. Any ideas? Funny this only happens for mobile pipelines.

@davidberard98 do you have any idea of what might be happening?

File: torch/csrc/jit/runtime/register_prim_ops.cpp
Comment (davidberard98) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:2296]: can you try calling this `"aten::device.with_index(str type, int index) -> Device"` and see if that fixes it?

Comment (thiagocrepaldi) [reply torch/csrc/jit/runtime/register_prim_ops.cpp]: good tip, just pushed a new revision, thanks!

My expectation is that all builds would fail because of that, no just mobile.

File: torch/csrc/jit/passes/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/peephole.cpp:279]: nit: null check for `toIValue(node->inputs().at(1))`

Comment (BowenBao) [inline torch/csrc/jit/passes/peephole.cpp:279]: nit
```suggestion
          int64_t index = toIValue(node->inputs().at(1))->toInt();
```

**Review by BowenBao — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@merge -g

**Comment by thiagocrepaldi** [comment]
@davidberard98 gentle ping

**Review by davidberard98 — APPROVED** [review]
I think @BowenBao has outstanding nits, otherwise lgtm!

**Pull request closed without merging**
